blockchain/iotex.go

```diff
@@ -0,0 +1,230 @@
+	conn   *iotexConnection
+	filter *iotexapi.LogsFilter
+}
+
+func (io *iotexSubscriber) SubscribeToEvents(channel chan<- subscriber.Event, _ ...interface{}) (subscriber.ISubscription, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	sub := io.newSubscription(channel, cancel, clock.New())
+	sub.run(ctx)
+	return sub, nil
+}
+
+func (io *iotexSubscriber) newSubscription(channel chan<- subscriber.Event, cancel context.CancelFunc, clk clock.Clock) *iotexSubscription {
+	return &iotexSubscription{
+		conn:         io.conn,
+		interval:     iotexScanInterval,
+		cancel:       cancel,
+		eventChannel: channel,
+		filter:       io.filter,
+		clock:        clk,
+	}
+}
+
+func (io *iotexSubscriber) Test() error {
+	if err := io.conn.connect(); err != nil {
+		return err
+	}
+	_, err := io.conn.api.GetChainMeta(context.Background(), &iotexapi.GetChainMetaRequest{})
+	return err
+}
+
+type iotexSubscription struct {
+	conn         *iotexConnection
+	interval     time.Duration
+	cancel       context.CancelFunc
+	eventChannel chan<- subscriber.Event
+	filter       *iotexapi.LogsFilter
+	clock        clock.Clock
+
+	ticker          *clock.Ticker
+	requestedHeight uint64
+}
+
+func (io *iotexSubscription) run(ctx context.Context) {
+	io.ticker = io.clock.Ticker(io.interval)
+	go func() {
+		for {
+			select {
+			case <-io.ticker.C:
+				io.poll(ctx)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (io *iotexSubscription) poll(ctx context.Context) {
+	if err := io.conn.connect(); err != nil {
+		logger.Error("failed to connect to iotex server:", err)
+		return
+	}
+	cm, err := io.conn.api.GetChainMeta(context.Background(), &iotexapi.GetChainMetaRequest{})
+	if err != nil {
+		logger.Error("failed to get iotex chain meta:", err)
+		return
+	}
+	currentHeight := cm.GetChainMeta().GetHeight()
+	fromHeight := currentHeight
+	count := uint64(1)
+	blocks := currentHeight - io.requestedHeight
+	if blocks == 0 {
+		return
+	} else if blocks < 1000 {
+		count = blocks
+		fromHeight = io.requestedHeight + 1
+	}
+	req := &iotexapi.GetLogsRequest{
+		Filter: io.filter,
+		Lookup: &iotexapi.GetLogsRequest_ByRange{
+			ByRange: &iotexapi.GetLogsByRange{
+				FromBlock: fromHeight,
+				Count:     count,
+			},
+		},
+	}
+	resp, err := io.conn.api.GetLogs(ctx, req)
+	if err != nil {
+		logger.Error("failed to get iotex event logs:", err)
+		return
+	}
+
+	// set sub height
+	io.requestedHeight = currentHeight
+
+	events, err := iotexLogEventToSubscriberEvents(resp.GetLogs())
+	if err != nil {
+		logger.Error("failed to convert iotex event logs to subscriber events:", err)
+		return
+	}
+
+	go func() {
+		for _, event := range events {
+			io.eventChannel <- event
+		}
+	}()
+}
+
+func createIoTeXLogFilter(jobid string, addresses []string) *iotexapi.LogsFilter {
+	topic := common.HexToHash(StringToBytes32(jobid))
+	return &iotexapi.LogsFilter{
+		Address: addresses,
+		Topics: []*iotexapi.Topics{
+			&iotexapi.Topics{
```

Comment:
```suggestion
			{
```
Redundant type declaration